scripts/performance/benchmark-local.sh

```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
+
+# This is intended solely to be called by scripts/performance/benchmark.sh. 
+# This file is separated out to allow us to easily execute this part of the test script inside a Docker container.
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+TEST_DOCS_FOLDER="$SCRIPT_DIR/docs"
+TIMEFORMAT="%R"
+
+function process_file() {
+    filepath=$1
+    strategy=$2
+    # access element in the return value to make sure we got something back, otherwise error
+    python3 -c 'from unstructured.partition.auto import partition; partition("'"$filepath"'", strategy="'"$strategy"'")[3]'
+}
+
+mkdir -p "$SCRIPT_DIR/benchmark_results" > /dev/null 2>&1
+DATE=$(date +"%Y-%m-%d_%H-%M-%S")
+RESULTS_FILE="$SCRIPT_DIR/benchmark_results/${DATE}_benchmark_results_${INSTANCE_TYPE}_$("$SCRIPT_DIR/get-stats-name.sh")_$GIT_HASH.csv"
+echo "Test File,Iterations,Average Execution Time (s)" > "$RESULTS_FILE"
+
+echo "Warming up..."
+process_file "$SCRIPT_DIR/warmup.pdf" "hi_res" > /dev/null
+
+echo "Starting benchmark test..."
+for file in "$TEST_DOCS_FOLDER"/*; do
+    echo "Testing file: $(basename "$file")"
+
+    if [[ " ${SLOW_FILES[*]} " =~ $(basename "$file") ]]; then
+        echo "File found in slow files list. Running once..."
+        num_iterations=1
+    else
+        # shellcheck disable=SC2153
+        num_iterations=$NUM_ITERATIONS
+    fi
+    total_execution_time=0
+    for ((i = 1; i <= num_iterations; i++)); do
+        echo "Iteration $i"
+        strategy="fast"
+        if [[ " ${HI_RES_STRATEGY_FILES[*]} " =~ $(basename "$file") ]]; then
+            echo "Testing with hi_res strategy"
+            strategy="hi_res"
+        fi
+        
+        if ! response=$( { time process_file "$file" "$strategy"; } 2>&1 ); then
```

Comment:
with this method, a new python process is spun up and many modules are imported as a result of calling `partition` (the warmup call above does force the issue for downloading needed modules, but there is still overhead for each fresh process after that).

the best way to reduce the noise is to make one call to `partition` call for the given filetype before measuring the next one, which implies doing more in the python execution side. E.g., the python script could return the time for the call after in-process warmup. The python process also could do the N iterations, rather doing one per process.

this may not be a terrible refactor, but it could also get punted to a future PR.

Reply:
yep, good call! getting this in. I think since we'll want to have some initial benchmarks to track progress against, worth having this in place now. from quick testing, module loading etc is clearly a a non-insignificant chunk of time spent.

Reply:
okay, should now have changes as per both of your recommendations. it will attempt to find an inexpensive warmup file of the same type or otherwise just use a pass on the test doc as the warm up. it also does all the iterations and timing inside of python.